[export][dynamo] Fallback to slowpath for MultiHeadAttention for strict export

anijain2305 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #164691
* __->__ #164721

In https://github.com/pytorch/pytorch/pull/106824, export decided to slow-path for MultiHeadAttention module (look into the PR description as to why). But that PR eventually caused a divergence between Dynamo and export. 

Today, strict-export does not inline into builtin modules (like MultiHeadAttention), and therefore make_fx sees the original nn.Module and takes the slow path. But compile inlines into the nn module, and at this time the condition `_is_make_fx_tracing` is False. As a result, Dynamo takes a fast path, resulting in a different op being called.


This divergence is undesirable. There are 2 ways to fix it

1) Make export take the fast path - As explained in the https://github.com/pytorch/pytorch/pull/106824 , this might be difficult. So, we go to (2)
2) Make compile as well take the slow path - This is easy to implement. The con here is that Pytorch eager and compile will use different operators, which can cause numerics issues etc.

Since (2) is easy to do, we will follow this path. We are tracking the issue in  https://github.com/pytorch/pytorch/issues/164062


cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @kadeng @chauhang @amjames @Lucaskabela

anijain2305 committed:
[export][dynamo] Fallback to slowpath for MultiHeadAttention for strict export

This is undesirable but MultiHeadAttention has some weirdness to it
where if we are make_fx tracing (_is_make_fx_tracing), it goes to slow
path. This weirdness leads to different behavior between non-strict
export and compile.

For non-strict export, _is_make_fx_tracing path is True and the captured
graph is slow path. But for compile, for Dynamo, the _is_make_fx_tracing
cond is False, leading to fast graph.

This inherent divergence leads to test failures in executorch. We should
figure out why _is_make_fx_tracing requries this separate condition.

[ghstack-poisoned]

anijain2305 committed:
Update on "[export][dynamo] Fallback to slowpath for MultiHeadAttention for strict export"


This is undesirable but MultiHeadAttention has some weirdness to it
where if we are make_fx tracing (_is_make_fx_tracing), it goes to slow
path. This weirdness leads to different behavior between non-strict
export and compile.

For non-strict export, _is_make_fx_tracing path is True and the captured
graph is slow path. But for compile, for Dynamo, the _is_make_fx_tracing
cond is False, leading to fast graph.

This inherent divergence leads to test failures in executorch. We should
figure out why _is_make_fx_tracing requries this separate condition.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames Lucaskabela

[ghstack-poisoned]

anijain2305 committed:
Update on "[export][dynamo] Fallback to slowpath for MultiHeadAttention for strict export"


This is undesirable but MultiHeadAttention has some weirdness to it
where if we are make_fx tracing (_is_make_fx_tracing), it goes to slow
path. This weirdness leads to different behavior between non-strict
export and compile.

For non-strict export, _is_make_fx_tracing path is True and the captured
graph is slow path. But for compile, for Dynamo, the _is_make_fx_tracing
cond is False, leading to fast graph.

This inherent divergence leads to test failures in executorch. We should
figure out why _is_make_fx_tracing requries this separate condition.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames Lucaskabela

[ghstack-poisoned]

avikchaudhuri commented on torch/nn/modules/activation.py:
```diff
@@ -1411,6 +1420,8 @@ def forward(
                 why_not_fast_path = "some Tensor argument has_torch_function"
             elif _is_make_fx_tracing():
                 why_not_fast_path = "we are running make_fx tracing"
+            elif _is_exporting():
```
...and move comment on `https://github.com/pytorch/pytorch/issues/164062` here

tugsbayasgalan commented on torch/nn/modules/activation.py:
```diff
@@ -1086,6 +1086,15 @@ def _is_make_fx_tracing():
+    # More details at https://github.com/pytorch/pytorch/issues/164062
+
+    # The weird code is because of TorchScript not returning early on `and` conditions.
+    if not torch.jit.is_scripting():
```
Yeah unfortunately TS doens't work in this case i think.

anijain2305 committed:
Update on "[export][dynamo] Fallback to slowpath for MultiHeadAttention for strict export"


This is undesirable but MultiHeadAttention has some weirdness to it
where if we are make_fx tracing (_is_make_fx_tracing), it goes to slow
path. This weirdness leads to different behavior between non-strict
export and compile.

For non-strict export, _is_make_fx_tracing path is True and the captured
graph is slow path. But for compile, for Dynamo, the _is_make_fx_tracing
cond is False, leading to fast graph.

This inherent divergence leads to test failures in executorch. We should
figure out why _is_make_fx_tracing requries this separate condition.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames Lucaskabela

[ghstack-poisoned]

anijain2305 committed:
Update on "[export][dynamo] Fallback to slowpath for MultiHeadAttention for strict export"


In https://github.com/pytorch/pytorch/pull/106824, export decided to slow-path for MultiHeadAttention module (look into the PR description as to why). But that PR eventually caused a divergence between Dynamo and export. 

Today, strict-export does not inline into builtin modules (like MultiHeadAttention), and therefore make_fx sees the original nn.Module and takes the slow path. But compile inlines into the nn module, and at this time the condition `_is_make_fx_tracing` is False. As a result, Dynamo takes a fast path, resulting in a different op being called.


This divergence is undesirable. There are 2 ways to fix it

1) Make export take the fast path - As explained in the https://github.com/pytorch/pytorch/pull/106824 , this might be difficult. So, we go to (2)
2) Make compile as well take the slow path - This is easy to implement. The con here is that Pytorch eager and compile will use different operators, which can cause numerics issues etc.

Since (2) is easy to do, we will follow this path. We are tracking the issue in  https://github.com/pytorch/pytorch/issues/164062


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames Lucaskabela

[ghstack-poisoned]
